// @partnerhero/phds.utilities.theme/dist/index.js
"use strict";
/**
 * Partner Hero Design System Theme definitions
 *  - Default Theme structure
 *    - Default Theme declaration
 *  - Declaration merging with styled-components
 *  - Color definitions
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFontContrastColor = exports.getDisabledColor = exports.getHoverColor = exports.getActiveColor = exports.KeywordColorEnum = exports.SemanticColorEnum = exports.PaletteColorEnum = exports.PlainColorsEnum = exports.ShadeEnum = exports.defaultTheme = void 0;
// Import the styled comp
require("./components/styled");
var theme_1 = require("./components/theme");
Object.defineProperty(exports, "defaultTheme", { enumerable: true, get: function () { return theme_1.defaultTheme; } });
var types_1 = require("./components/types");
Object.defineProperty(exports, "ShadeEnum", { enumerable: true, get: function () { return types_1.ShadeEnum; } });
Object.defineProperty(exports, "PlainColorsEnum", { enumerable: true, get: function () { return types_1.PlainColorsEnum; } });
Object.defineProperty(exports, "PaletteColorEnum", { enumerable: true, get: function () { return types_1.PaletteColorEnum; } });
Object.defineProperty(exports, "SemanticColorEnum", { enumerable: true, get: function () { return types_1.SemanticColorEnum; } });
Object.defineProperty(exports, "KeywordColorEnum", { enumerable: true, get: function () { return types_1.KeywordColorEnum; } });
var utils_1 = require("./components/utils");
Object.defineProperty(exports, "getActiveColor", { enumerable: true, get: function () { return utils_1.getActiveColor; } });
Object.defineProperty(exports, "getHoverColor", { enumerable: true, get: function () { return utils_1.getHoverColor; } });
Object.defineProperty(exports, "getDisabledColor", { enumerable: true, get: function () { return utils_1.getDisabledColor; } });
Object.defineProperty(exports, "getFontContrastColor", { enumerable: true, get: function () { return utils_1.getFontContrastColor; } });
