// @partnerhero/phds.utilities.theme/dist/components/types.js
"use strict";
// Definitions in this file are only for values that a theme can have, it does
// NOT define the colors themselves.
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeywordColorEnum = exports.PlainColorsEnum = exports.ShadeEnum = exports.PaletteColorEnum = exports.SemanticColorEnum = void 0;
/**
 * SEMANTIC colors for the PHDS. Will have different shades
 *
 * If you need more, you can use declaration merging to extend this object.
 *
 * *Notes:*
 * - Whenever **refactoring** a color name, please make sure to alter both the key and the value
 */
var SemanticColorEnum;
(function (SemanticColorEnum) {
    SemanticColorEnum["primary"] = "primary";
    SemanticColorEnum["secondary"] = "secondary";
    SemanticColorEnum["success"] = "success";
    SemanticColorEnum["notice"] = "notice";
    SemanticColorEnum["error"] = "error";
})(SemanticColorEnum = exports.SemanticColorEnum || (exports.SemanticColorEnum = {}));
/**
 * PALETTE colors for the PHDS. Will have different shades.
 *
 * If you need more, you can use declaration merging to extend this object.
 *
 * *Notes:*
 * - Whenever **refactoring** a color name, please make sure to alter both the `key` and the `value`
 */
var PaletteColorEnum;
(function (PaletteColorEnum) {
    PaletteColorEnum["grey"] = "grey";
})(PaletteColorEnum = exports.PaletteColorEnum || (exports.PaletteColorEnum = {}));
/**
 * SHADES for the PHDS.
 *
 * If you need more, you can use declaration merging
 * to extend this object.
 *
 * *Notes:*
 * - Declare new shades in PascalCase as they will be composed together with the semantic
 * colors to create a camelCase color name.
 * - Whenever **refactoring** a color name, please make sure to alter both the key and the value
 */
var ShadeEnum;
(function (ShadeEnum) {
    ShadeEnum["Darkest"] = "Darkest";
    ShadeEnum["Darker"] = "Darker";
    ShadeEnum["Lighter"] = "Lighter";
    ShadeEnum["Lightest"] = "Lightest";
})(ShadeEnum = exports.ShadeEnum || (exports.ShadeEnum = {}));
/**
 * PLAIN colors (No shades!) for the PHDS.
 * For now also used for edge case colors. Like background colors.
 * Commented colors means that definitions were not found when migrating to V2.
 *
 * If you need more, you can use declaration merging to extend this object.
 */
var PlainColorsEnum;
(function (PlainColorsEnum) {
    PlainColorsEnum["defaultColor"] = "defaultColor";
    PlainColorsEnum["black"] = "black";
    PlainColorsEnum["white"] = "white";
    PlainColorsEnum["primaryBackground"] = "primaryBackground";
    PlainColorsEnum["secondaryBackground"] = "secondaryBackground";
    PlainColorsEnum["mediaPlaceholderGrey"] = "mediaPlaceholderGrey";
    // backgroundPrimaryLightest = 'backgroundPrimaryLightest',
    // backgroundSecondaryLightest = 'backgroundSecondaryLightest',
    // neutral = 'neutral',
    // highlight = 'primary',
    // info = 'primary'
})(PlainColorsEnum = exports.PlainColorsEnum || (exports.PlainColorsEnum = {}));
/**
 * Contains accepted css keywords that are not rgb colors.
 * @example 'inherit'
 */
var KeywordColorEnum;
(function (KeywordColorEnum) {
    KeywordColorEnum["transparent"] = "transparent";
    KeywordColorEnum["inherit"] = "inherit";
})(KeywordColorEnum = exports.KeywordColorEnum || (exports.KeywordColorEnum = {}));
