// @partnerhero/phds.utilities.theme/components/styled.ts
// import original module declarations
import 'styled-components';
import Color from 'color';
import { Colors, KeywordColor } from './types';

/**
 * An object that maps the keyword colors to their values.
 * @example { 'inherit': 'inherit'}
 */
type KeywordColorDefs = {
    [key in KeywordColor]: string;
};

/**
 * Only color names, does not contain keyword colors.
 */
export type RgbColorNames = Exclude<Colors, KeywordColor>;

/**
 * An object that maps color names to their `Color` values.
 */
type RgbColorDefs = {
    [key in RgbColorNames]: Color;
};

/**
 * The different value types that Design System colors can be.
 */
export type ColorValues = Color | string;

/**
 * Merged color definitions that include both rgb colors and css keyword colors
 */
export type ColorDefinitions = KeywordColorDefs & RgbColorDefs;

/**
 * The color theme for the PHDS.
 * To be declaration merged with the default theme from styled-components.
 *
 * If you require any more properties, please add them here.
 */
export interface DSTheme {
    colors: ColorDefinitions;
}

declare module 'styled-components' {
    export interface DefaultTheme extends DSTheme {}
}
