// @partnerhero/phds.utilities.typography/dist/typography.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Typography = exports.TypographyFontAwesomeSizeOptions = exports.TypographyOptions = exports.boldExtraSmall = exports.boldSmall = exports.boldMedium = exports.boldLarge = exports.defaultBoldStyles = exports.bodyExtraSmall = exports.bodySmall = exports.bodyMedium = exports.bodyLarge = exports.header5 = exports.header4 = exports.header3 = exports.header2 = exports.header1 = exports.defaultHeaderStyles = exports.defaultNavigationStyles = exports.remToValue = exports.pxToRem = void 0;
/* eslint-disable no-console */
const react_1 = __importDefault(require("react"));
const styled_components_1 = __importStar(require("styled-components"));
const phds_utilities_theme_1 = require("@partnerhero/phds.utilities.theme");
/**
 * Converts px to rem using a base font size of 16px
 * rounded to 4 decimal places so that when multiplied by the fontRoot, it will give a whole number
 *
 * example: `pxToRem(16);` ==> `16px -> 1rem`
 * @param px number of px to convert
 * @returns a string with the value in rem
 */
const pxToRem = (px, fontRoot = 16) => {
    const rem = (px / fontRoot).toFixed(4);
    const value = parseFloat(rem);
    return `${value}rem`;
};
exports.pxToRem = pxToRem;
/**
 * Extracts the rem's float value
 *
 * example: `remToValue('16rem');` ==> 16
 * @param remValue rem value from which its pure float will be extracted
 * @returns a number representing the rem's float value
 */
const remToValue = (remValue) => remValue ? Number(remValue.replace('rem', '')) : 0;
exports.remToValue = remToValue;
const fontSizes = {
    h1: (0, exports.pxToRem)(45),
    h2: (0, exports.pxToRem)(35),
    h3: (0, exports.pxToRem)(30),
    h4: (0, exports.pxToRem)(25),
    h5: (0, exports.pxToRem)(18),
    large: (0, exports.pxToRem)(16),
    medium: (0, exports.pxToRem)(14),
    small: (0, exports.pxToRem)(12),
    xsmall: (0, exports.pxToRem)(10),
};
const fontMargins = {
    h1: `0 0 ${(0, exports.pxToRem)(25)} 0`,
    h2: `0 0 ${(0, exports.pxToRem)(20)} 0`,
    h3: `0 0 ${(0, exports.pxToRem)(15)} 0`,
    h4: `0 0 ${(0, exports.pxToRem)(15)} 0`,
    h5: `0 0 ${(0, exports.pxToRem)(10)} 0`,
    large: `0 0 ${(0, exports.pxToRem)(35)} 0`,
    medium: `0 0 ${(0, exports.pxToRem)(25)} 0`,
    small: `0 0 ${(0, exports.pxToRem)(20)} 0`,
    xsmall: `0 0 ${(0, exports.pxToRem)(15)} 0`,
};
const variantDefinitions = {
    h1: (0, styled_components_1.css) `
        font-size: ${fontSizes.h1};
        ${(props) => (0, styled_components_1.css) `
            margin: ${props.margin ? `${props.margin}` : fontMargins.h1};
        `};
    `,
    h2: (0, styled_components_1.css) `
        font-size: ${fontSizes.h2};
        ${(props) => (0, styled_components_1.css) `
            margin: ${props.margin ? `${props.margin}` : fontMargins.h2};
        `};
    `,
    h3: (0, styled_components_1.css) `
        font-size: ${fontSizes.h3};
        ${(props) => (0, styled_components_1.css) `
            margin: ${props.margin ? `${props.margin}` : fontMargins.h3};
        `};
    `,
    h4: (0, styled_components_1.css) `
        font-size: ${fontSizes.h4};
        ${(props) => (0, styled_components_1.css) `
            margin: ${props.margin ? `${props.margin}` : fontMargins.h4};
        `};
    `,
    h5: (0, styled_components_1.css) `
        font-size: ${fontSizes.h5};
        ${(props) => (0, styled_components_1.css) `
            margin: ${props.margin ? `${props.margin}` : fontMargins.h5};
        `};
    `,
    large: (0, styled_components_1.css) `
        font-size: ${fontSizes.large};
        ${(props) => (0, styled_components_1.css) `
            margin: ${props.margin ? `${props.margin}` : fontMargins.large};
        `};
    `,
    medium: (0, styled_components_1.css) `
        font-size: ${fontSizes.medium};
        ${(props) => (0, styled_components_1.css) `
            margin: ${props.margin ? `${props.margin}` : fontMargins.medium};
        `};
    `,
    small: (0, styled_components_1.css) `
        font-size: ${fontSizes.small};
        ${(props) => (0, styled_components_1.css) `
            margin: ${props.margin ? `${props.margin}` : fontMargins.small};
        `};
    `,
    xsmall: (0, styled_components_1.css) `
        font-size: ${fontSizes.xsmall};
        ${(props) => (0, styled_components_1.css) `
            margin: ${props.margin ? `${props.margin}` : fontMargins.xsmall};
        `};
    `,
};
const defaultTypographyStyles = (0, styled_components_1.css) `
    letter-spacing: 0;
    display: flex;
    align-items: center;
    justify-content: ${(props) => props.align ? `${props.align}` : 'left'};
`;
// NAVIGATION
exports.defaultNavigationStyles = (0, styled_components_1.css) `
    ${defaultTypographyStyles}
    font-family: Inter-SemiBold;
    font-weight: 600;
    font-size: ${(0, exports.pxToRem)(12)};
    line-height: ${(0, exports.pxToRem)(26)};
`;
// HEADERS
exports.defaultHeaderStyles = (0, styled_components_1.css) `
    ${defaultTypographyStyles}
    font-family: Inter-SemiBold;
    line-height: normal;
    font-weight: 600;
`;
exports.header1 = (0, styled_components_1.css) `
    ${exports.defaultHeaderStyles}
    ${variantDefinitions.h1}
`;
exports.header2 = (0, styled_components_1.css) `
    ${exports.defaultHeaderStyles}
    ${variantDefinitions.h2}
`;
exports.header3 = (0, styled_components_1.css) `
    ${exports.defaultHeaderStyles}
    ${variantDefinitions.h3}
`;
exports.header4 = (0, styled_components_1.css) `
    ${exports.defaultHeaderStyles}
    ${variantDefinitions.h4}
`;
exports.header5 = (0, styled_components_1.css) `
    ${exports.defaultHeaderStyles}
    ${variantDefinitions.h5}
`;
// BODY
/**
 * Default body styles
 */
const defaultBodyStyles = (0, styled_components_1.css) `
    ${defaultTypographyStyles}
    font-family: Inter-Regular;
    font-weight: normal;
`;
/**
 * 16 px (when root font size is 16px)
 */
exports.bodyLarge = (0, styled_components_1.css) `
    ${defaultBodyStyles}
    ${variantDefinitions.large}
    line-height: ${(0, exports.pxToRem)(26)};
`;
/**
 * 14 px (when root font size is 16px)
 */
exports.bodyMedium = (0, styled_components_1.css) `
    ${defaultBodyStyles}
    ${variantDefinitions.medium}
    line-height: ${(0, exports.pxToRem)(22)};
`;
/**
 * 12 px (when root font size is 16px)
 */
exports.bodySmall = (0, styled_components_1.css) `
    ${defaultBodyStyles}
    ${variantDefinitions.small}
    line-height: ${(0, exports.pxToRem)(20)};
`;
/**
 * 10 px (when root font size is 16px)
 */
exports.bodyExtraSmall = (0, styled_components_1.css) `
    ${defaultBodyStyles}
    ${variantDefinitions.xsmall}
    line-height: ${(0, exports.pxToRem)(20)};
`;
// BOLD
exports.defaultBoldStyles = (0, styled_components_1.css) `
    ${defaultTypographyStyles}
    font-family: Inter-SemiBold;
    font-weight: 600;
    line-height: ${(0, exports.pxToRem)(22)};
`;
exports.boldLarge = (0, styled_components_1.css) `
    ${exports.defaultBoldStyles}
    ${variantDefinitions.large}
`;
exports.boldMedium = (0, styled_components_1.css) `
    ${exports.defaultBoldStyles}
    ${variantDefinitions.medium}
`;
exports.boldSmall = (0, styled_components_1.css) `
    ${exports.defaultBoldStyles}
    ${variantDefinitions.small}
`;
exports.boldExtraSmall = (0, styled_components_1.css) `
    ${exports.defaultBoldStyles}
    ${variantDefinitions.xsmall}
`;
/**
 *  To add new typography styles, add them to this object.
 */
exports.TypographyOptions = {
    /**
     * The font size for the `h1` tag
     * @default '45px' in rem
     */
    h1: exports.header1,
    /**
     * The font size for the `h2` tag
     * @default '35px' in rem
     */
    h2: exports.header2,
    /**
     * The font size for the `h3` tag
     * @default '30px' in rem
     */
    h3: exports.header3,
    /**
     * The font size for the `h4` tag
     * @default '25px' in rem
     */
    h4: exports.header4,
    /**
     * The font size for the `h5` tag
     * @default '18px' in rem
     */
    h5: exports.header5,
    /**
     * The font size for the `bodyLarge` and `boldLarge` tag
     * @default '16px' in rem
     */
    boldLarge: exports.boldLarge,
    /**
     * The font size for the `bodyMedium` and `boldMedium` tag
     * @default '14px' in rem
     */
    boldMedium: exports.boldMedium,
    /**
     * The font size for the `bodySmall` and `boldSmall` tag
     * @default '12px' in rem
     */
    boldSmall: exports.boldSmall,
    /**
     * The font size for the `bodyExtraSmall` and `boldExtraSmall` tag
     * @default '10px' in rem
     */
    boldExtraSmall: exports.boldExtraSmall,
    /**
     * The font size for the `bodyLarge` and `boldLarge` tag
     * @default '16px' in rem
     */
    bodyLarge: exports.bodyLarge,
    /**
     * The font size for the `bodyMedium` and `boldMedium` tag
     * @default '14px' in rem
     */
    bodyMedium: exports.bodyMedium,
    /**
     * The font size for the `bodySmall` and `boldSmall` tag
     * @default '12px' in rem
     */
    bodySmall: exports.bodySmall,
    /**
     * The font size for the `bodyExtraSmall` and `boldExtraSmall` tag
     * @default '10px' in rem
     */
    bodyExtraSmall: exports.bodyExtraSmall,
};
exports.TypographyFontAwesomeSizeOptions = {
    xs: (0, exports.pxToRem)(12),
    sm: (0, exports.pxToRem)(14),
    lg: (0, exports.pxToRem)(20),
    '24px': (0, exports.pxToRem)(24),
    '1x': (0, exports.pxToRem)(16),
    '2x': (0, exports.pxToRem)(32),
    '3x': (0, exports.pxToRem)(48),
    '4x': (0, exports.pxToRem)(64),
    '5x': (0, exports.pxToRem)(80),
    '6x': (0, exports.pxToRem)(96),
    '7x': (0, exports.pxToRem)(112),
    '8x': (0, exports.pxToRem)(128),
    '9x': (0, exports.pxToRem)(144),
    '10x': (0, exports.pxToRem)(160),
};
const TypographyBody = styled_components_1.default.div `
    color: ${(props) => `${props.theme.colors[props.color]}`};
    ${(props) => props.margin &&
    (0, styled_components_1.css) `
            margin: ${props.margin};
        `};
    ${(props) => props.align &&
    (0, styled_components_1.css) `
            justify-content: ${props.align};
        `};
    ${(props) => exports.TypographyOptions[props.variant]};
    ${({ customVariant }) => customVariant};
`;
function Typography(_a) {
    var { children, color = 'inherit', margin, variant = 'bodyMedium', align = 'left', style, customVariant } = _a, rest = __rest(_a, ["children", "color", "margin", "variant", "align", "style", "customVariant"]);
    const typographyTag = (exports.TypographyOptions[variant] === exports.header1 && 'h1') ||
        (exports.TypographyOptions[variant] === exports.header2 && 'h2') ||
        (exports.TypographyOptions[variant] === exports.header3 && 'h3') ||
        (exports.TypographyOptions[variant] === exports.header4 && 'h4') ||
        (exports.TypographyOptions[variant] === exports.header5 && 'h5') ||
        (exports.TypographyOptions[variant] === exports.boldExtraSmall && 'strong') ||
        (exports.TypographyOptions[variant] === exports.boldSmall && 'strong') ||
        (exports.TypographyOptions[variant] === exports.boldMedium && 'strong') ||
        (exports.TypographyOptions[variant] === exports.boldLarge && 'strong') ||
        'p';
    return (react_1.default.createElement(styled_components_1.ThemeProvider, { theme: phds_utilities_theme_1.defaultTheme },
        react_1.default.createElement(TypographyBody, Object.assign({ as: typographyTag, color: color, margin: margin, align: align, variant: variant, style: style, customVariant: customVariant }, rest), children)));
}
exports.Typography = Typography;
